src/config_format/flb_cf_yaml.c

```diff
@@ -1338,114 +1408,252 @@ static char *get_real_path(char *file, char *path, size_t size)
     return path;
 }
 
-static void state_destroy(struct parser_state *s)
+static struct parser_state *state_start(struct local_ctx *ctx, struct file_state *file)
+{
+    struct parser_state *state;
+
+    state = state_create(NULL, file);
+    if (state != NULL) {
+        cfl_list_add(&state->_head, &ctx->states);
+    }
+
+    return state;
+}
+
+static struct parser_state *state_push(struct local_ctx *ctx, enum state state_num)
 {
-    if (s->caller_file) {
-        flb_sds_destroy(s->caller_file);
+    struct parser_state *last = NULL;
+    struct parser_state *state;
+
+    if (cfl_list_size(&ctx->states) <= 0) {
+        return NULL;
     }
-    if (s->caller_root_path) {
-        flb_sds_destroy(s->caller_root_path);
+
+    last = cfl_list_entry_last(&ctx->states, struct parser_state, _head);
+    state = state_create(last->file, last->file);
+    if (state == NULL) {
+        return NULL;
     }
+    state->section = last->section;
+    state->keyvals = last->keyvals;
+    state->cf_section = last->cf_section;
+    state->cf_group = last->cf_group;
+    state->values = last->values;
+    state->file = last->file;
+    state->state = state_num;
+    state->level = last->level + 1;
+    state->key = last->key;
+
+    cfl_list_add(&state->_head, &ctx->states);
+    return state;
+}
 
-    if (s->file) {
-        flb_sds_destroy(s->file);
+static struct parser_state *state_push_section(struct local_ctx *ctx,
+                                               enum state state_num,
+                                               enum section section)
+{
+    struct parser_state *state;
+
+    state = state_push(ctx, state_num);
+    if (state == NULL) {
+        return NULL;
     }
+    state->section = section;
 
-    if (s->root_path) {
-        flb_sds_destroy(s->root_path);
+    return state;
+}
+
+static struct parser_state *state_push_key(struct local_ctx *ctx,
+                                           enum state state_num,
+                                           const char *key)
+{
+    struct parser_state *state;
+    flb_sds_t skey;
+
+    skey = flb_sds_create(key);
```

Comment:
If `key` is `NULL`  then `flb_sds_create` will create and empty string, is that acceptable or an error?